From 6ba40cd480d5ca19142fc040480f270246dd43bc Mon Sep 17 00:00:00 2001
From: mj <mj@maxlinear.com>
Date: Tue, 26 Aug 2025 20:25:38 -0700
Subject: [PATCH] WLANRTSYS-92970 get channel to freq based on the second best
 channel index Issue: second best freq is not valid which we get from ACS
 table. Fix: ACS table always update lower freq for each channel. So,no need
 to take freq from the acs table. convert secondary channel to freq and update
 second best freq for channel switch after radar detected.

---
 src/ap/mxl_acs_proprietary.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/src/ap/mxl_acs_proprietary.c b/src/ap/mxl_acs_proprietary.c
index 405a6e0db..aaa8201ed 100644
--- a/src/ap/mxl_acs_proprietary.c
+++ b/src/ap/mxl_acs_proprietary.c
@@ -4610,6 +4610,8 @@ bool mxl_get_second_best_channel_from_acs_table(struct hostapd_iface *iface)
 	struct csa_settings settings;
 	struct hostapd_channel_data *channel = NULL;
 	int secondary_channel;
+	int sec_best_freq;
+	int sec_op_class;
 	u8 vht_oper_centr_freq_seg0_idx;
 	u8 vht_oper_centr_freq_seg1_idx;
 	struct acs_candidate_table *candidate = iface->mxl_iface.candidates;
@@ -4619,12 +4621,14 @@ bool mxl_get_second_best_channel_from_acs_table(struct hostapd_iface *iface)
 	channel = mxl_dfs_get_channel_dynamic_failsafe(iface, &secondary_channel, &vht_oper_centr_freq_seg0_idx, &vht_oper_centr_freq_seg1_idx, false, &settings.freq_params);
 
 	if (channel) {
+		hostapd_get_op_class(candidate[iface->mxl_iface.sec_best_ch.second_rank_idx].primary, candidate[iface->mxl_iface.sec_best_ch.second_rank_idx].width, &sec_op_class);
+		sec_best_freq = ieee80211_chan_to_freq(NULL, sec_op_class, candidate[iface->mxl_iface.sec_best_ch.second_rank_idx].primary);
 		iface->mxl_iface.sec_best_ch.channel = candidate[iface->mxl_iface.sec_best_ch.second_rank_idx].chan;
-		iface->mxl_iface.sec_best_ch.freq = candidate[iface->mxl_iface.sec_best_ch.second_rank_idx].freq;
+		iface->mxl_iface.sec_best_ch.freq = sec_best_freq;
 		iface->mxl_iface.sec_best_ch.width = candidate[iface->mxl_iface.sec_best_ch.second_rank_idx].width;
 		iface->mxl_iface.sec_best_ch.primary = candidate[iface->mxl_iface.sec_best_ch.second_rank_idx].primary;
 		iface->mxl_iface.sec_best_ch.secondary = candidate[iface->mxl_iface.sec_best_ch.second_rank_idx].secondary;
-		wpa_printf(MSG_DEBUG, "ACS: second rank_cand_idx %d,  chan : %d", iface->mxl_iface.sec_best_ch.second_rank_idx, iface->mxl_iface.sec_best_ch.primary);
+		wpa_printf(MSG_DEBUG, "ACS: second rank_cand_idx %d,  chan : %d, freq: %d", iface->mxl_iface.sec_best_ch.second_rank_idx, iface->mxl_iface.sec_best_ch.primary, iface->mxl_iface.sec_best_ch.freq);
 		return true;
 	} else {
 		return false;
-- 
2.43.0

